src/codec.rs

```diff
@@ -488,29 +550,125 @@ impl<T, X> Encode for X where
 pub trait WrapperTypeDecode: Sized {
 	/// A wrapped type.
 	type Wrapped: Into<Self>;
+
+	// !INTERNAL USE ONLY!
+	// This is a used to specialize `decode` for the wrapped type.
+	#[doc(hidden)]
+	fn decode_wrapped<I: Input>(_input: &mut I) -> Option<Result<Self, Error>> where Self::Wrapped: Decode {
+		// Not implemented by default, so `Wrapped::decode` will be called instead.
+		None
+	}
 }
+
 impl<T> WrapperTypeDecode for Box<T> {
 	type Wrapped = T;
+
+	fn decode_wrapped<I: Input>(input: &mut I) -> Option<Result<Self, Error>> where Self::Wrapped: Decode {
+		if let Err(error) = input.descend_ref() {
+			return Some(Err(error));
+		}
+
+		// Placement new is not yet stable, but we can just manually allocate a chunk of memory
+		// and convert it to a `Box` ourselves.
+		//
+		// The explicit types here are written out for clarity.
+		//
+		// TODO: Use `Box::new_uninit` once that's stable.
+		let layout = core::alloc::Layout::from_size_align(
+			core::mem::size_of::<MaybeUninit<T>>(),
+			core::mem::align_of::<MaybeUninit<T>>()
+		).expect("layout is always valid");
```

Comment:
Can we just do this?
```suggestion
        let layout = Layout::new::<MaybeUninit<T>>();
```
Source: https://doc.rust-lang.org/src/core/alloc/layout.rs.html#148-154